packages/openid4vc-client/src/OpenId4VcClientService.ts

```diff
@@ -1,10 +1,201 @@
+      if (!jwt.header) {
+        throw new AriesFrameworkError('No header present on JWT')
+      }
+
+      if (!jwt.payload) {
+        throw new AriesFrameworkError('No payload present on JWT')
+      }
+      if (!kid.startsWith('did:')) {
+        throw new AriesFrameworkError(`kid '${kid}' is not a valid did. Only dids are supported as kid.`)
+      }
+
+      if (!kid.includes('#')) {
+        throw new AriesFrameworkError(
+          `kid '${kid}' does not include a reference to the verificationMethod. The kid must specify a specific verificationMethod within the did document .`
+        )
+      }
+
+      const did = kid.split('#')[0]
+
+      const didsApi = agentContext.dependencyManager.resolve(DidsApi)
+      const [didRecord] = await didsApi.getCreatedDids({ did })
+
+      if (!didRecord) {
+        throw new AriesFrameworkError(`No did record found for did ${did}. Is the did created by this agent?`)
+      }
+
+      const didResult = await didsApi.resolve(did)
+
+      if (!didResult.didDocument) {
+        throw new AriesFrameworkError(
+          `No did document found for did ${did}. ${didResult.didResolutionMetadata.error} - ${didResult.didResolutionMetadata.message}`
+        )
+      }
+
+      // TODO: which purposes are allowed?
+      const verificationMethod = didResult.didDocument.dereferenceKey(kid, ['authentication'])
+      const { getKeyFromVerificationMethod } = getKeyDidMappingByVerificationMethod(verificationMethod)
+      const key = getKeyFromVerificationMethod(verificationMethod)
+
+      const payload = JsonEncoder.toBuffer(jwt.payload)
+
+      const jws = await this.jwsService.createJwsCompact(agentContext, {
+        key, // FIXME null check
+        payload,
+        protectedHeaderOptions: {
+          alg: jwt.header.alg,
+          kid: jwt.header.kid,
+        },
+      })
+
+      return jws
+    }
+  }
+
+  private getSignCallback(agentContext: AgentContext) {
+    return {
+      signCallback: this.signCallback(agentContext),
+    }
+  }
+
+  public async preAuthorized(
+    agentContext: AgentContext,
+    options: PreAuthorizedOptions,
+    checkRevocationState = true
+  ): Promise<W3cCredentialRecord> {
+    this.logger.debug('Running pre-authorized flow with options', options)
+
+    // The clientId is set to a dummy value for now as we don't have
+    // one in the pre-auth flow, but it's currently required by OpenID4VCIClient.
+    const clientId = 'some-client'
+
+    const client = await OpenID4VCIClient.initiateFromURI({
+      issuanceInitiationURI: options.issuerUri,
+      flowType: AuthzFlowType.PRE_AUTHORIZED_CODE_FLOW,
+      kid: options.kid,
+      alg: Alg.EdDSA,
+      clientId: clientId,
+    })
+
+    const accessToken = await client.acquireAccessToken({ clientId: clientId })
+
+    this.logger.info('Fetched server accessToken', accessToken)
+
+    // We currently need the ts-ignore because the type
+    // inside of OpenID4VCIClient needs to be updated.
+    // @ts-ignore
```

Comment:
Is there an issue/PR in the client lib?

Reply:
I've fixed this, and it has been merged. The client lib doesn't have an automatic release pipeline though, so it hasn't been released yet.

Reply:
👍